events/entity_counts.go

```diff
@@ -0,0 +1,97 @@
+package events
+
+import (
+	"fmt"
+	"github.com/openziti/edge/controller/persistence"
+	"github.com/openziti/foundation/util/cowslice"
+	"github.com/pkg/errors"
+	"go.etcd.io/bbolt"
+	"reflect"
+	"time"
+)
+
+const EntityCountEventNS = "edge.entityCounts"
+
+type EntityCountEvent struct {
+	Namespace string           `json:"namespace"`
+	Timestamp time.Time        `json:"timestamp"`
+	Counts    map[string]int64 `json:"counts"`
+	Error     string           `json:"error"`
+}
+
+func (event *EntityCountEvent) String() string {
+	return fmt.Sprintf("%v timestamp=%v counts=%+v err=%v",
+		event.Namespace, event.Timestamp, event.Counts, event.Error)
+}
+
+var entityCountEventGenerator func(interval time.Duration)
+
+func generateEntityCountEvent(dbProvider persistence.DbProvider, stores *persistence.Stores) {
+	handlers := getEntityCountEventHandlers()
+	if len(handlers) == 0 {
+		return
+	}
+
+	event := &EntityCountEvent{
+		Namespace: EntityCountEventNS,
+		Timestamp: time.Now(),
+		Counts:    map[string]int64{},
+	}
+
+	for _, store := range stores.GetStoreList() {
+		err := dbProvider.GetDb().View(func(tx *bbolt.Tx) error {
+			_, count, err := store.QueryIds(tx, "true limit 1")
+			if err != nil {
+				return err
+			}
+			event.Counts[store.GetEntityType()] = count
+			return nil
+		})
+
+		if err != nil {
+			event.Error = err.Error()
+			break
+		}
+	}
+
+	for _, handler := range handlers {
+		handler.AcceptEntityCountEvent(event)
+	}
+}
+
+type EntityCountEventHandler interface {
+	AcceptEntityCountEvent(event *EntityCountEvent)
+}
+
+var entityCountEventHandlerRegistry = cowslice.NewCowSlice(make([]EntityCountEventHandler, 0))
+
+func getEntityCountEventHandlers() []EntityCountEventHandler {
+	return entityCountEventHandlerRegistry.Value().([]EntityCountEventHandler)
+}
+
+func registerEntityCountEventHandler(val interface{}, config map[interface{}]interface{}) error {
+	handler, ok := val.(EntityCountEventHandler)
+
+	if !ok {
+		return errors.Errorf("type %v doesn't implement github.com/openziti/edge/events/EntityCountEventHandler interface.", reflect.TypeOf(val))
+	}
+
+	interval := time.Minute * 5
+
+	if val, ok := config["interval"]; ok {
+		if strVal, ok := val.(string); ok {
+			var err error
+			interval, err = time.ParseDuration(strVal)
+			if err != nil {
+				return errors.Wrapf(err, "invalid duration value for edge.entityCounts interval: '%v'", strVal)
+			}
+		} else {
+			return errors.Errorf("invalid type %v for edge.entityCounts interval configuration", reflect.TypeOf(val))
+		}
+	}
+
+	go entityCountEventGenerator(interval)
```

Comment:
This function calls `AddEntityCountEventHandler` which adds to `entityCountEventHandlerRegistry`. 

Then  `go entityCountEventGenerator`  calls `generateEntityCountEvent` which in turn emits events for all handlers by calling `getEntityCountEventHandlers` which comes from `entityCountEventHandlerRegistry. 

Doesn't this mean that as `registerEntityCountEventHandler` is called multiple times every configured interval will trigger every other handler?

Reply:
Not any more :) 

I had considered this, but with the typical configuration, it wouldn't be a problem since we'd only ever have a single handler. But it was bothering me as well, so I fixed it. This means I removed the Add/Remove handlers, so this is a bit different from the other event types. However, if someone is in-process, they don't need this is event, they can query for it directly, so I think that's fine.

Reply:
Nice, good update.